WordPress/openverse
# Content moderator user preferences admin view

## Fixes



Fixes #3634 by @AetherUnbound

## Description



This PR adds a new `UserPreferences` ["profile" model](https://docs.djangoproject.com/en/5.0/topics/auth/customizing/#extending-user) to the project, along with a new Django Admin view for editing a user's preferences. The admin view only ever displays the currently logged in user's preferences, and it has a custom form set up to set (or unset) a specific value within the `jsonb` column used for storing preferences.

In a bit of a follow up to #3633, I've also added some startup logic to create both an additional `moderator` user with staff access to the Django Admin site and a "Content Moderators" group which has the permissions we're expecting for that group. The upshot of this is that all of the testing below shouldn't require any additional user or group creation! Yay!

This was really tricky to figure out how to match the expectations of the IP while doing all this within the Django framework. I haven't done a _ton_ of Django work, so if there are more idiomatic ways to accomplish what I've done, please suggest them!

## Screenshots

**Moderator user view**
![image](https://github.com/WordPress/openverse/assets/10214785/fb9d3a30-c92d-4123-aa09-82a1480503f7)

**Moderator preferences edit form**
![image](https://github.com/WordPress/openverse/assets/10214785/43b92014-dec9-44b2-b82b-56f71605a3a3)


## Testing Instructions


I added some tests specifically for the admin UI form - those should pass (and hopefully make sense).

1. Run `just api/dbshell`, then execute `select u.username, p.preferences from auth_user u left join api_userpreferences p on p.user_id = u.id;`. There should be 3 users and all preferences should be blank.
2. Log in to the admin site (http://localhost:50280/admin/) using `deploy`/`deploy`. Observe the new "USER PREFERENCES" section with the "My Preferences" entry.
3. Click on "My Preferences". Observe that only `deploy's preferences` shows up in the list.
4. Click on `deploy's preferences`. Observe that the form only includes a check box for "Blur images".
5. Log out and log in again using `moderator`/`deploy`. Observe that only the report/sensitivity tables and user preferences sections show up.
6. Follow the steps above to edit the user preferences (verifying only the moderator user shows up), only this time unchecking the "Blur images" box. Click save.
7. Run the select above once more. The `preferences` value for the `moderator` user should be `{"moderator": {"blur_images": false}}`
8. Edit the preferences once more time and re-enable blurring images. Run the select above and verify the preferences saved in the database are now `{"moderator": {"blur_images": true}}`


## Checklist



[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
